lib/blockscore/base.rb

```diff
@@ -0,0 +1,59 @@
+require 'active_support/core_ext/class'
+require 'active_support/core_ext/string/inflections'
+require 'blockscore/connection'
+
+module BlockScore
+  class Base
+    extend BlockScore::Connection
+
+    class_attribute :api_key, :resource, :version
+
+    def self.inherited(base)
+      base.resource = base.to_s.split('::').last.underscore
+    end
+
+    def initialize(options = {})
+      options[:class] = resource
+      @attrs = options
+    end
+
+    def refresh
+      self.class.retrieve(id)
+      instance_variable_set(:@attrs, r.instance_variable_get(:@attrs))
+
+      true
+    rescue
+      false
+    end
+
+    def self.auth(api_key)
+      api_key = api_key
+      BlockScore::Connection.api_key = api_key
+    end
+
+    def self.api_url
+      'https://api.blockscore.com/'
+    end
+
+    def self.endpoint
+      if self == BlockScore::Base
+        fail NotImplementedError.new('Base is an abstract class. You should ' +
```

Comment:
We usually prefer the syntax

```ruby
fail SomeError, 'message'
```

over creating an exception instance. Just a style guide thing https://github.com/bbatsov/ruby-style-guide#exception-class-messages